src/Nethermind/Nethermind.Optimism/CL/P2P/OptimismCLP2P.cs

```diff
@@ -0,0 +1,227 @@
+    private PubsubRouter? _router;
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+    private readonly ILogger _logger;
+    private readonly IOptimismEngineRpcModule _engineRpcModule;
+    private readonly IP2PBlockValidator _blockValidator;
+    private readonly Multiaddress[] _staticPeerList;
+    private readonly ICLConfig _config;
+    private ILocalPeer? _localPeer;
+
+    private readonly string _blocksV2TopicId;
+
+    private ITopic? _blocksV2Topic;
+
+    private const int MaxGossipSize = 10485760;
+
+    public OptimismCLP2P(ulong chainId, string[] staticPeerList, ICLConfig config, Address sequencerP2PAddress, ITimestamper timestamper, ILogManager logManager, IOptimismEngineRpcModule engineRpcModule)
+    {
+        _logger = logManager.GetClassLogger();
+        _config = config;
+        _staticPeerList = staticPeerList.Select(addr => Multiaddress.Decode(addr)).ToArray();
+        _engineRpcModule = engineRpcModule;
+        _blockValidator = new P2PBlockValidator(chainId, sequencerP2PAddress, timestamper, _logger);
+
+        _blocksV2TopicId = $"/optimism/{chainId}/2/blocks";
+
+        _serviceProvider = new ServiceCollection()
+            .AddSingleton<PeerStore>()
+            .AddLibp2p(builder => builder)
+            .AddSingleton(new IdentifyProtocolSettings
+            {
+                ProtocolVersion = "",
+                AgentVersion = "optimism"
+            })
+            .AddSingleton(new Settings())
+            .BuildServiceProvider();
+    }
+
+    private ulong _headPayloadNumber = 0;
+    private readonly SemaphoreSlim _semaphore = new(1);
+    private async void OnMessage(byte[] msg)
+    {
+        await _semaphore.WaitAsync();
+        try
+        {
+            if (TryValidateAndDecodePayload(msg, out var payload))
+            {
+                if (_logger.IsTrace) _logger.Trace($"Received payload prom p2p: {payload}");
+
+                if (_headPayloadNumber >= (ulong)payload.BlockNumber)
+                {
+                    // Old payload. skip
+                    return;
+                }
+
+                if (await SendNewPayloadToEL(payload) && await SendForkChoiceUpdatedToEL(payload.BlockHash))
+                {
+                    _headPayloadNumber = (ulong)payload.BlockNumber;
+                }
+            }
+        }
+        finally
+        {
+            _semaphore.Release();
+        }
+    }
+
+    private bool TryValidateAndDecodePayload(byte[] msg, [MaybeNullWhen(false)] out ExecutionPayloadV3 payload)
+    {
+        int length = Snappy.GetUncompressedLength(msg);
+        if (length < 65 || length > MaxGossipSize)
+        {
+            payload = null;
+            return false;
+        }
+
+        byte[] decompressed = new byte[length];
+        Snappy.Decompress(msg, decompressed);
+
+        byte[] signature = decompressed[0..65];
+        byte[] payloadData = decompressed[65..];
```

Comment:
use spans instead of allocating arrays?

Reply:
Reply for the comment bellow: Payloads are ssz encoded, so data offsets are encoded and they are relative to the beginning. We potentially can calculate total prefix offset and subtract it from encoded offsets, but I don't think it's good idea.